# Exploring Hacker News Posts

---

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

In [8]:
print(hn[0:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [9]:
headers = hn[0]

In [10]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [11]:
hn = hn[1:]

In [12]:
hn[0:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Separate `Ask` and `Show` to `Other` posts

In [14]:
ask_posts = list()
show_posts = list()
other_posts = list()
for row in hn:
    title = row[1]
    lowered = title.lower()
    if lowered.startswith("ask hn"):
        ask_posts.append(row)
    elif lowered.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [15]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## Average Number of Comments for `Ask` and `Show` posts

In [16]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [17]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


Clearly, `Ask` posts receive more comments on average at 14.038, compared to `Show` posts at 10.317.

## Calculating average number of comments for `Ask` posts by Hour

In [18]:
import datetime as dt
result_list = list()

In [19]:
for row in ask_posts:
    dt_obj = row[6]
    num_comments = int(row[4])
    result_list.append([dt_obj, num_comments])

In [20]:
counts_by_hour = dict()
comments_by_hour = dict()

In [21]:
dt_format = "%m/%d/%Y %H:%M"
for each in result_list:
    comment = int(each[1])
    dt_obj = each[0]
    date_time = dt.datetime.strptime(dt_obj, dt_format)
    hora = date_time.hour
    hour = int(hora)
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment

In [22]:
counts_by_hour

{0: 55,
 1: 60,
 2: 58,
 3: 54,
 4: 47,
 5: 46,
 6: 44,
 7: 34,
 8: 48,
 9: 45,
 10: 59,
 11: 58,
 12: 73,
 13: 85,
 14: 107,
 15: 116,
 16: 108,
 17: 100,
 18: 109,
 19: 110,
 20: 80,
 21: 109,
 22: 71,
 23: 68}

In [23]:
comments_by_hour

{0: 447,
 1: 683,
 2: 1381,
 3: 421,
 4: 337,
 5: 464,
 6: 397,
 7: 267,
 8: 492,
 9: 251,
 10: 793,
 11: 641,
 12: 687,
 13: 1253,
 14: 1416,
 15: 4477,
 16: 1814,
 17: 1146,
 18: 1439,
 19: 1188,
 20: 1722,
 21: 1745,
 22: 479,
 23: 543}

In [25]:
avg_by_hour = list()
for hr in counts_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
avg_by_hour

[[0, 8.127272727272727],
 [1, 11.383333333333333],
 [2, 23.810344827586206],
 [3, 7.796296296296297],
 [4, 7.170212765957447],
 [5, 10.08695652173913],
 [6, 9.022727272727273],
 [7, 7.852941176470588],
 [8, 10.25],
 [9, 5.5777777777777775],
 [10, 13.440677966101696],
 [11, 11.051724137931034],
 [12, 9.41095890410959],
 [13, 14.741176470588234],
 [14, 13.233644859813085],
 [15, 38.5948275862069],
 [16, 16.796296296296298],
 [17, 11.46],
 [18, 13.20183486238532],
 [19, 10.8],
 [20, 21.525],
 [21, 16.009174311926607],
 [22, 6.746478873239437],
 [23, 7.985294117647059]]

In [26]:
swap_avg_by_hour = list()
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
swap_avg_by_hour

[[8.127272727272727, 0],
 [11.383333333333333, 1],
 [23.810344827586206, 2],
 [7.796296296296297, 3],
 [7.170212765957447, 4],
 [10.08695652173913, 5],
 [9.022727272727273, 6],
 [7.852941176470588, 7],
 [10.25, 8],
 [5.5777777777777775, 9],
 [13.440677966101696, 10],
 [11.051724137931034, 11],
 [9.41095890410959, 12],
 [14.741176470588234, 13],
 [13.233644859813085, 14],
 [38.5948275862069, 15],
 [16.796296296296298, 16],
 [11.46, 17],
 [13.20183486238532, 18],
 [10.8, 19],
 [21.525, 20],
 [16.009174311926607, 21],
 [6.746478873239437, 22],
 [7.985294117647059, 23]]

In [28]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [29]:
sorted_swap

[[38.5948275862069, 15],
 [23.810344827586206, 2],
 [21.525, 20],
 [16.796296296296298, 16],
 [16.009174311926607, 21],
 [14.741176470588234, 13],
 [13.440677966101696, 10],
 [13.233644859813085, 14],
 [13.20183486238532, 18],
 [11.46, 17],
 [11.383333333333333, 1],
 [11.051724137931034, 11],
 [10.8, 19],
 [10.25, 8],
 [10.08695652173913, 5],
 [9.41095890410959, 12],
 [9.022727272727273, 6],
 [8.127272727272727, 0],
 [7.985294117647059, 23],
 [7.852941176470588, 7],
 [7.796296296296297, 3],
 [7.170212765957447, 4],
 [6.746478873239437, 22],
 [5.5777777777777775, 9]]

In [31]:
print("Top 5 Hours for Ask Posts Comments")
for each in sorted_swap[0:5]:
    print("{a}:00: {b:.2f} average comments per post".format(a=each[1], b=each[0]))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
2:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
